apps/opik-backend/src/main/java/com/comet/opik/infrastructure/bi/InstallationReportService.java

```diff
@@ -0,0 +1,135 @@
+import reactor.core.scheduler.Schedulers;
+import ru.vyarus.dropwizard.guice.module.lifecycle.GuiceyLifecycle;
+
+import java.net.URI;
+import java.util.Map;
+import java.util.Optional;
+
+@ImplementedBy(InstallationReportServiceImpl.class)
+interface InstallationReportService {
+    String NOTIFICATION_EVENT_TYPE = "opik_os_startup_be";
+
+    void reportInstallation();
+}
+
+@Singleton
+@Slf4j
+@RequiredArgsConstructor(onConstructor_ = @Inject)
+class InstallationReportServiceImpl implements InstallationReportService {
+
+    private final LockService lockService;
+    private final IdGenerator generator;
+    private final UsageReportService usageReport;
+    private final OpikConfiguration config;
+    private final Client client;
+
+    public void reportInstallation() {
+
+        String eventType = GuiceyLifecycle.ApplicationStarted.name();
+
+        if (!config.getUsageReport().isEnabled()) {
+            log.info("Usage report is disabled");
+            return;
+        }
+
+        if (StringUtils.isEmpty(config.getUsageReport().getUrl())) {
+            log.warn("Usage report URL is not set");
+            return;
+        }
+
+        var lock = new LockService.Lock("opik-%s".formatted(eventType));
+
+        lockService.executeWithLock(lock, tryToReportStartupEvent(eventType))
+                .subscribeOn(Schedulers.boundedElastic())
+                .onErrorResume(e -> {
+                    log.warn("Didn't reported due to error", e);
+                    return Mono.empty();
+                }).block();
+
+    }
+
+    private Mono<Void> tryToReportStartupEvent(String eventType) {
+        return Mono.fromCallable(() -> {
+
+            String anonymousId = getAnonymousId();
+
+            log.info("Anonymous ID: {}", anonymousId);
+
+            if (usageReport.isEventReported(eventType)) {
+                log.info("Event already reported");
+                return null;
+            }
+
+            reportEvent(anonymousId, eventType);
+            return null;
+        });
+    }
+
+    private String getAnonymousId() {
+        var anonymousId = usageReport.getAnonymousId();
+
+        if (anonymousId.isEmpty()) {
+            log.info("Anonymous ID not found, generating a new one");
+            var newId = generator.generateId();
+            log.info("Generated new ID: {}", newId);
+
+            // Save the new ID
+            usageReport.saveAnonymousId(newId.toString());
+
+            anonymousId = Optional.of(newId.toString());
+        }
+
+        return anonymousId.get();
+    }
+
+    private void reportEvent(String anonymousId, String eventType) {
+
+        usageReport.addEvent(eventType);
+
+        var startupEvent = new OpikStartupEvent(
+                anonymousId,
+                NOTIFICATION_EVENT_TYPE,
+                Map.of("opik_app_version", config.getMetadata().getVersion()));
+
+        try (Response response = client.target(URI.create(config.getUsageReport().getUrl()))
```

Comment:
Please double check what thread is really executing the event listener callback. 

Otherwise as you end-up blocking the reactive call, we need to ensure that service never hangs at start-up time.

This applies to the HTTP client and any other client such as Redis, MySQL etc.

Reply:
This should be avoided by the `Schedulers.boundedElastic()`. But let me add a log to demonstrate it.

Reply:
```
INFO  [2024-09-30 12:10:13,865] com.comet.opik.infrastructure.bi.InstallationReportServiceImpl: Thread name: main
DEBUG [2024-09-30 12:10:13,880] com.comet.opik.infrastructure.redis.RedissonLockService: Trying to lock with Lock[key=opik-ApplicationStarted]
DEBUG [2024-09-30 12:10:13,927] com.comet.opik.infrastructure.redis.RedissonLockService: Lock Lock[key=opik-ApplicationStarted] acquired
INFO  [2024-09-30 12:10:13,928] com.comet.opik.infrastructure.bi.InstallationReportServiceImpl: Thread name: boundedElastic-2
INFO  [2024-09-30 12:10:13,960] com.comet.opik.infrastructure.bi.InstallationReportServiceImpl: Anonymous ID not found, generating a new one
INFO  [2024-09-30 12:10:13,960] com.comet.opik.infrastructure.bi.InstallationReportServiceImpl: Generated new ID: efada259-f543-445a-9473-82c9f3334ad6
INFO  [2024-09-30 12:10:13,971] com.comet.opik.infrastructure.bi.InstallationReportServiceImpl: Thread name: boundedElastic-2
INFO  [2024-09-30 12:10:13,971] com.comet.opik.infrastructure.bi.InstallationReportServiceImpl: Anonymous ID: efada259-f543-445a-9473-82c9f3334ad6
INFO  [2024-09-30 12:10:13,979] com.comet.opik.infrastructure.bi.InstallationReportServiceImpl: Thread name: boundedElastic-2
INFO  [2024-09-30 12:10:14,115] com.comet.opik.infrastructure.bi.InstallationReportServiceImpl: Event reported successfully: Event added successfully
DEBUG [2024-09-30 12:10:14,116] com.comet.opik.infrastructure.redis.RedissonLockService: Lock Lock[key=opik-ApplicationStarted] released
```